arcgis-ios-sdk-samples/Utility Network/Configure subnetwork trace/ConfigureSubnetworkTraceViewController.swift

```diff
@@ -0,0 +1,389 @@
+            }
+            optionsViewController.title = "Attributes"
+            show(optionsViewController, sender: self)
+        } else if cell == comparisonCell {
+            // Prompt comparison operator selection.
+            let optionsViewController = OptionsTableViewController(labels: comparisonsStrings, selectedIndex: comparisonsStrings.count) { (index) in
+                self.selectedComparison = self.comparisons[index]
+                self.comparisonLabel?.text = self.comparisonsStrings[index]
+            }
+            optionsViewController.title = "Comparison"
+            show(optionsViewController, sender: self)
+        } else if cell == valueCell {
+            if selectedAttribute != nil {
+                if let domain = selectedAttribute?.domain as? AGSCodedValueDomain {
+                    // Get the value labels.
+                    if valueLabels.isEmpty {
+                        domain.codedValues.forEach { (codedValue) in
+                            valueLabels.append(codedValue.name)
+                        }
+                    }
+                    // Prompt value selection.
+                    let optionsViewController = OptionsTableViewController(labels: valueLabels, selectedIndex: domain.codedValues.count) { (index) in
+                        self.valueLabel?.text = self.valueLabels[index]
+                        self.selectedValue = domain.codedValues[index]
+                    }
+                    optionsViewController.title = "Value"
+                    show(optionsViewController, sender: self)
+                } else {
+                    // Prompt the user to create a custom value if none are available to select.
+                    addCustomValue()
+                }
+            } else {
+                // If no attribute is selected, present an alert.
+                presentAlert(title: "No attribute selected", message: "Please select an attribute before selecting a value.")
+            }
+        } else if cell == addConditionButton {
+            addCondition()
+        } else if cell == resetButton {
+            reset()
+        } else if cell == traceButton {
+            trace()
+        }
+    }
+    
+    func addCustomValue() {
+        // Create an alert controller.
+        let alert = UIAlertController(title: "Create a value", message: "This attribute has no values. Please create one.", preferredStyle: .alert)
+        // Add the text field.
+        alert.addTextField { (textField) in
+            textField.placeholder = "Add a custom value"
+            textField.keyboardType = .numberPad
+        }
+        // Obtain user value.
+        alert.addAction(UIAlertAction(title: "Done", style: .default) { [weak self] _ in
+            let textfield = alert.textFields![0]
+            // Check for invalid input characters.
+            if !(CharacterSet(charactersIn: "0123456789").isSuperset(of: CharacterSet(charactersIn: textfield.text!))) {
```

Comment:
Whenever possible, text field input validation should be done on-demand rather than when the user taps the "Done" button. That is normally done through observing changes to the text field's text via its delegate, particularly by implementing `UITextFieldDelegate.textField(_:shouldChangeCharactersIn:replacementString:)`. Does that work with a text field in an alert? I don't know that I've ever tried to validate text in an alert's text field, but I don't see why that wouldn't work. For what it is worth, the best way I've found to validate numeric input is using a number formatter to try to create a number from a given string. There's just a couple edge cases when dealing with decimal numbers.